garden_ai/local_data.py

```diff
@@ -149,6 +150,164 @@ def _delete_old_model_versions(model_name: str):
+                if r_data[f] is None:
+                    resource_table_row.append("None")
+                else:
+                    resource_table_row.append(r_data[f])
+            resource_table_rows.append(tuple(resource_table_row))
+        for col in resource_table_cols:
+            table.add_column(col)
+        for row in resource_table_rows:
+            table.add_row(*(row))
+    else:
+        for col in resource_table_cols:
+            table.add_column(col)
+
+    return table
+
+
+def _get_local_resource_json(id_: str, resource_type: ResourceType) -> Optional[Dict]:
+    data = _read_local_db()
+    if "/" in id_ and resource_type is not ResourceType.MODEL:
+        resources_by_uuid = data.get(resource_type.value, {})
+        resources_by_doi = _reindex_by_doi(resources_by_uuid)
+        if resources_by_doi and id_ in resources_by_doi:
+            return resources_by_doi[id_]
+        else:
+            return None
+    else:
+        id_ = str(id_)
+        resources = data.get(resource_type.value, {})
+        if resources and id_ in resources:
+            return resources[id_]
+        else:
+            return None
+
+
+def get_local_garden_table(fields: List[str], table_name: str) -> Table:
+    """Helper: fetch all gardens from ~/.garden/data.json and outputs a rich table to print.
+
+    Parameters
+    ----------
+    fields List
+        A list of the garden fields you want included as cols in the table.
+    table_name str
+        Name of the rich table
+
+    Returns
+    -------
+    Table
+        Returns a rich table of all local gardens.
+    """
+    return _get_local_resource_table(
+        resource_type=ResourceType.GARDEN, fields=fields, table_name=table_name
+    )
+
+
+def get_local_pipeline_table(fields: List[str], table_name: str) -> Table:
+    """Helper: fetch all pipelines from ~/.garden/data.json and outputs a rich table to print.
+
+    Parameters
+    ----------
+    fields List
+        A list of the pipeline fields you want included as cols in the table.
+    table_name str
+        Name of the rich table
+
+    Returns
+    -------
+    Table
+        Returns a rich table of all local pipelines.
+    """
+    return _get_local_resource_table(
+        resource_type=ResourceType.PIPELINE, fields=fields, table_name=table_name
+    )
+
+
+def get_local_model_table(fields: List[str], table_name: str) -> Table:
+    """Helper: fetch all models from ~/.garden/data.json and outputs a rich table to print.
+
+    Parameters
+    ----------
+    fields List
+        A list of the model fields you want included as cols in the table.
+    table_name str
+        Name of the rich table
+
+    Returns
+    -------
+    Table
+        Returns a rich table of all local models.
+    """
+    return _get_local_resource_table(
+        resource_type=ResourceType.MODEL, fields=fields, table_name=table_name
+    )
+
+
+def get_local_garden_json(garden_id: Union[UUID, str]) -> Optional[Dict]:
```

Comment:
(sorry for the wall of text) this is totally not obvious and something I probably could gone into earlier, but since this is kinda the reason we're using pydantic models/dataclasses for everything in the first place I'm jumping on the opportunity to elaborate:

This module shouldn't have public functions that put or get dicts/json directly. We want to stick to functions that put or get a complete instance of one of our dataclasses, so that other modules don't need to think about whether or not the json/dict they'd be using is valid. 

For example: consider the difference between the dict I'd get from your function, 
```
data1: Dict = get_local_garden_json("<uuid-bunchanumbers>")
``` 
vs from 
```
obj = get_local_garden_by_uuid("<uuid-bunchanumbers>")
data2: Dict = obj.dict()
```
The former would be a bit faster, because it returns the dictionary element directly instead of instantiating an object in the process. But what if the json it came from wasn't valid, e.g. the entry didn't have a `"doi"` key? 

In the latter case, whatever code uses that `data2` dictionary never needs to worry about whether it's a valid record -- without a `"doi"` key, `obj` could not have instantiated, so `data2['doi']` never needs to worry about KeyErrors, while `data1['doi']` does.

Reply:
that is, pydantic makes it natural to validate all your data before/when you instantiate (usually called parsing elsewhere, but pydantic calls them validators 🤷‍♂️)